### DISCUSSION TITLE
Matching parallel types while keeping typescript happy

### DISCUSSION THREAD

**Discussion by p-j** [statelyai/xstate#1919] [Q&A] [answered]
Hi,

I'm running into an issue that I feel is more related to typescript than to xstate but maybe someone has workaround I could take inspiration of...

Here is a simplified version of the code I'm using: https://codesandbox.io/s/kind-turing-cu4xf?file=/src/index.ts

While the parallel states are being executed, the state value of the machine will be (in the case of the example):

```ts
{ parallel: { one: 'pending', two: 'pending', three: 'pending' } }
// after 1 sec
{ parallel: { one: 'success', two: 'pending', three: 'pending' } }
// after 5 sec
{ parallel: { one: 'success', two: 'success', three: 'pending' } }
// after 30 sec
{ parallel: { one: 'success', two: 'success', three: 'success' } }
```

How can I match the state value
```ts
{ parallel: { one: 'success', two: 'success', three: any } }
```
while keeping Typescript happy?

Currently I have tried (as implemented in the example shared above)

```ts
['parallel.one.success', 'parallel.two.success'].every((substate) => state.matches(substate))
// TS complains: Argument of type 'string' is not assignable to parameter of type '"init" | "parallel" | "parallel.one.pending" | "parallel.one.success" | "parallel.one.error" | "parallel.two.pending" | "parallel.two.success" | "parallel.two.error" | "parallel.three.pending" | "parallel.three.success" | "parallel.three.error" | "end"'

state.value.parallel.one === 'success' && state.value.parallel.two === 'success'
// TS complains: Property 'parallel' does not exist on type 'StateValue'.
//               Property 'parallel' does not exist on type 'string'.

state.matches('parallel.one.success') && state.matches('parallel.two.success')
// This one looks good at first... until you realize that within the "if" state is now typed as "never"...
```

Although TS complains, running as is will execute correctly for all 3 "solutions".

Has anyone found a way of matching on parallel state while keeping TS happy?
Do I need to modify my TypeState somehow?

Thanks

**Comment by Andarist** [comment, marked as answer]
You can do this:
```js
if (
  ([
    "parallel.one.success",
    "parallel.two.success",
  ] as const).every((substate) => state.matches(substate))
) {
  console.log("Every match you make...", state);
}
```

It won't narrow your `state` down based on typestates though. For that, we'd have to figure out how to make smth like this work in TS:
```js
if (
  state.matches({
    parallel: {
      one: "success",
      two: "success",
    },
  })
) {
  console.log("I'll be watching you...", state);
}

```

**Reply by p-j** [reply]
Thanks, the `as const` did the trick for me, although it didn't work in the code sandbox, it worked in the actual codebase. Probably a TypeScript version issue.

And indeed, I was first looking for something like your second proposal, but I can see why it would make things more complicated to use it like that.

It sure would be better to have typed narrowed down to what is actually listed in the TypeState but for now that'll do.

Thanks for your quick response @Andarist
